thoth/storages/graph/postgres.py

```diff
@@ -1213,7 +1215,117 @@ def _create_python_software_stack(
+                run_cpu = OpenShift.parse_cpu_spec(document["specification"]["run"]["requests"]["cpu"])
+                run_memory = OpenShift.parse_memory_spec(document["specification"]["run"]["requests"]["memory"])
+
+                # Convert bytes to GiB, we need float number given the fixed int size.
+                run_memory = run_memory / (1024 ** 3)
+                build_memory = build_memory / (1024 ** 3)
+
+                run_hardware_information, run_software_environment = self._runtime_environment_conf2models(
+                    document["specification"]["run"].get("requests", {}), environment_type="RUNTIME", is_user=False
+                )
+
+                build_hardware_information, build_software_environment = self._runtime_environment_conf2models(
+                    document["specification"]["build"].get("requests", {}),
+                    environment_type="BUILDTIME",
+                    is_user=False,
+                )
+
+                if "python" in document["specification"]:
+                    # Inspection stack.
+                    software_stack = self._create_python_software_stack(
+                        software_stack_type="INSPECTION",
+                        requirements=document["specification"].get("requirements"),
+                        requirements_lock=document["specification"].get("requirements_locked"),
+                        software_environment=run_software_environment,
+                        performance_score=None,
+                        overall_score=None,
+                    )
+
+                inspection_run = (
+                    self._session.query(InspectionRun)
+                    .filter(InspectionRun.inspection_document_id == inspection_document_id)
+                    .first()
+                )
+
+                if inspection_run and inspection_run.dependency_monkey_run_id:
+                    # If inspection was run through Depedency Monkey
+                    inspection_run = InspectionRun(
+                        inspection_sync_state="SYNCED",
+                        inspection_document_id=inspection_document_id,
+                        datetime=document.get("created"),
+                        amun_version=None,  # TODO: propagate Amun version here which should match API version
+                        build_requests_cpu=build_cpu,
+                        build_requests_memory=build_memory,
+                        run_requests_cpu=run_cpu,
+                        run_requests_memory=run_memory,
+                        build_software_environment_id=build_software_environment.id,
+                        build_hardware_information_id=build_hardware_information.id,
+                        run_software_environment_id=run_software_environment.id,
+                        run_hardware_information_id=run_hardware_information.id,
+                    )
+                    self._session.upsert(inspection_run)
```

Comment:
Does it provide `upsert`?

Anyway, I think this will not work as expected - it will create a new version which will eventually fail on `inspection_document_id` unique constraint.

I think there needs to be queried model first (just on `inspection_document_id`) and behave in two cases - dependency monkey already created it or we are creating it.

Reply:
Actually there is https://docs.sqlalchemy.org/en/13/dialects/postgresql.html?highlight=conflict#insert-on-conflict-upsert